### ISSUE TITLE
No results returned on JOIN query

### ISSUE THREAD

**Issue by faymarie** [crate/crate#14003]
### CrateDB version

5.2.6

### CrateDB setup information

Number of nodes: 3
Heap: 4G

### Problem description

After upgrading CrateDB from `4.2.7 `-> `5.2.6` one of our queries returns zero results when it should return results.

The issue appeared 24h after the upgrade!

This is the simplified query, which returns no results anymore after the upgrade:

```
SELECT * FROM table_a JOIN table_b ON  table_b.x = 'some-string' 
AND table_b.y = table_a.z;
```

We managed to fix the issue by swapping the `JOIN` conditions:

```
SELECT * FROM table_a JOIN table_b ON  table_b.y = table_a.z 
AND table_b.x = 'some-string';
```

The `EXPLAIN` statement of the original query does reveal no difference in joining the tabes in both CrateDB versions:

`EXPLAIN SELECT * FROM config_nodes JOIN config_node_connections ON config_node_connections.type = 'used.in' AND config_node_connections.c_in = config_nodes.id limit 100;`

**Result for CrateDB 4.2.7:**

--> Returns results

```
Fetch[d, id, name, parent, template, auth_graph, name_search, c_in, c_out, idx, meta, type]
  └ Limit[100::bigint;0]
    └ HashJoin[((type = 'used.in') AND (c_in = id))]
      ├ Collect[doc.config_nodes | [_fetchid, id] | true]
      └ Collect[doc.config_node_connections | [_fetchid, c_in, type] | true]
```

**Result for CrateDB 5.2.6:**

--> 0 results returned

```
Fetch[d, id, name, parent, template, auth_graph, name_search, c_in, c_out, idx, meta, type]
  └ Limit[100::bigint;0]
    └ HashJoin[((type = 'used.in') AND (c_in = id))]
      ├ Collect[doc.config_nodes | [_fetchid, id] | true]
      └ Collect[doc.config_node_connections | [_fetchid, c_in, type] | true]
```

Not sure, if this is relevant, but for completeness, I would like to share the `EXPLAIN ` results with the join conditions swapped. There is a visible difference, but we get the same results for both.

```
EXPLAIN SELECT * FROM config_nodes JOIN config_node_connections ON config_node_connections.c_in = config_nodes.id AND config_node_connections.type = 'used.in' limit 100;
```

**Result for CrateDB 4.2.7:**

--> Notice the INNER NestedLoopJoin here:

```
Fetch[d, id, name, parent, template, auth_graph, name_search, c_in, c_out, idx, meta, type]
  └ Limit[100::bigint;0]
    └ NestedLoopJoin[INNER | ((c_in = id) AND (type = 'used.in'))]
      ├ Collect[doc.config_nodes | [_fetchid, id] | true]
      └ Collect[doc.config_node_connections | [_fetchid, c_in, type] | true]
```

**Result for CrateDB 5.2.6:**

```
Fetch[d, id, name, parent, template, auth_graph, name_search, c_in, c_out, idx, meta, type]
  └ Limit[100::bigint;0]
    └ HashJoin[(c_in = id)]
      ├ Collect[doc.config_nodes | [_fetchid, id] | true]
      └ Collect[doc.config_node_connections | [_fetchid, c_in] | (type = 'used.in')]
```

### Steps to Reproduce

See above.

### Actual Result

```
SELECT * FROM table_a JOIN table_b ON  table_b.x = 'some-string' 
AND table_b.y = table_a.z;
```

Returns zero results.

### Expected Result

Returns results.

**Comment by matriv** [comment]
With the data shared privately, (I only imported the `id` column for `config_nodes`, I get total 104 rows from the join, and 100 if I add the limit.

**Comment by faymarie** [comment]
@matriv Please be aware that the issue appears only after 24h have passed.

**Comment by matriv** [comment]
I know, and I also ran analyze to get the stats, but in this case we have a hash join and not a nestedloop, which can be affected by the stats (swapped tables).

**Comment by matriv** [comment]
I've managed to reproduce the issue and found the bug, meanwhile as a workaround please split the join condition and pass the filter to a where clause like this:
```
SELECT * FROM config_nodes JOIN config_node_connections ON config_node_connections.c_in = config_nodes.id WHERE config_node_connections.type = 'used.in' limit 20; 
```

**Comment by matriv** [comment]
The issue seems to be that after getting statistics, the tables are swapped, and the symbols used for the hash calculation of the left relation (former right) are more than the ones required, and I think the issue was introduced with: https://github.com/crate/crate/pull/12460/

**Comment by matriv** [comment]
The issue seems to be that after getting statistics, the tables are swapped, and the symbols used for the hash calculation of the left relation (former right) are more than the ones required, and I think the issue was introduced with: https://github.com/crate/crate/pull/12460/

**Comment by matriv** [comment]
The issue is fixed from 5.3.0 onwards with: https://github.com/crate/crate/pull/13802,

So the PR opened addresses the issue for 5.2

**Comment by matriv** [comment]
@faymarie Thx a lot for reporting this bug, the fix has been merged for 5.2 and will be available with the next hotfix release.
